File: 01/part-2.php
<?php
$input = file_get_contents(__DIR__ . "/input.txt");
$input = explode("\n", $input);

// Separate the input into left and right ids
$l_ids = [];
$r_ids = [];
foreach ($input as $line) {
	[$l, $r] = explode("   ", $line);
	$l_ids[]= (int) $l;
	$r_ids[]= (int) $r;
}

// Create a id => count map
$id_count = array_count_values($r_ids);

// Calculate the similarity score
$score = array_reduce($l_ids, function ($carry, $l) use ($id_count) {
    return $carry + $l * ($id_count[$l] ?? 0);
}, 0);

// Print the score
print_r($score);

File: 01/part-2.c
#include <stdio.h>
#include <stdlib.h>

#define INPUT_FILE "input.txt"
#define LOCATION_PAIRS 1000

int main() {
  int score = 0;
  int l[LOCATION_PAIRS], r[LOCATION_PAIRS];
  int counts[99999] = {0};

  FILE *input = fopen(INPUT_FILE, "r");
  if (input == NULL) {
    printf("Error: '%s' not found\n", INPUT_FILE);
    return 1;
  }

  // Parse input into two arrays of ints
  for (size_t i = 0; i < LOCATION_PAIRS; i++) {
    fscanf(input, "%d   %d\n", &l[i], &r[i]);
  }

  fclose(input);

  // Create count lookup for the locations in the right array
  for (size_t i = 0; i < LOCATION_PAIRS; i++) {
    counts[r[i]] += 1;
  }

  // Calculate the score
  for (size_t i = 0; i < LOCATION_PAIRS; i++) {
    int loc = l[i];
    int count = counts[loc];
    score += loc * count;
  }

  printf("%d\n", score);

  return 0;
}
